# How to Use YelpAPI:

In [1]:
# Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import json
with open('/Users/tijes/.secret/yelp_api.json') as f:
    login = json.load(f)
login.keys()

dict_keys(['client-id', 'api-key'])

In [3]:
# import the YelpAPI Class
from yelpapi import YelpAPI
# Create an instance with your key
yelp_api = YelpAPI(login['api-key'], timeout_s=5.0)
yelp_api

In [4]:
help(yelp_api.search_query)

Help on method search_query in module yelpapi.yelpapi:

search_query(**kwargs) method of yelpapi.yelpapi.YelpAPI instance
    Query the Yelp Search API.
    
    documentation: https://www.yelp.com/developers/documentation/v3/business_search
    
    required parameters:
        * one of either:
            * location - text specifying a location to search for
            * latitude and longitude



In [5]:
# use our yelp_api variable's search_query method to perform our API call
search_results = yelp_api.search_query(location='Baltimore, MD,21202',
                                       term='Crab Cakes')
print(type(search_results))
search_results.keys()

<class 'dict'>


dict_keys(['businesses', 'total', 'region'])

In [6]:
search_results['total']

434

It looks like there were 434 total results for this Yelp search.

In [7]:
biz = pd.DataFrame(search_results['businesses'])
biz.head(2)

id                                 alias  \
0  D9A33FM394q99o4QtK5YwA          faidleys-seafood-baltimore-3   
1  ieS_5zqxDHcWMCm8BKUYbg  thames-street-oyster-house-baltimore   

                         name  \
0           Faidley's Seafood   
1  Thames Street Oyster House   

                                           image_url  is_closed  \
0  https://s3-media3.fl.yelpcdn.com/bphoto/OTjVDC...      False   
1  https://s3-media1.fl.yelpcdn.com/bphoto/9hGjo5...      False   

                                                 url  review_count  \
0  https://www.yelp.com/biz/faidleys-seafood-balt...          1189   
1  https://www.yelp.com/biz/thames-street-oyster-...          2736   

                                          categories  rating  \
0  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.0   
1  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.5   

                                       coordinates transactions price  \
0  {'latitude': 39.291696, 'longitude': -76.62224}   [delivery]    $$   
1   {'latitude': 39.28214, 'longitude': -76.59162}   [delivery]   $$$   

                                            location         phone  \
0  {'address1': '203 N Paca St', 'address2': '', ...  +14107274898   
1  {'address1': '1728 Thames St', 'address2': '',...  +14434497726   

    display_phone     distance  
0  (410) 727-4898  1349.560720  
1  (443) 449-7726  2090.712792

In [8]:
len(search_results['businesses'])

20

It looks like there are 20 businesses stored in the list.

# Pagination and "offset"

Pagination can be understood through this brief demonstration:

<i> Total Results vs Businesses

If we check the total# of results, we can see that we had over 400 businesses that met our search criteria.

In [9]:
## total number of matching businesses
search_results['total']

434

However, if we check our list of businesses, we will see that we only received data for 20 businesses.

In [10]:
## how many businesses in our results
len(search_results['businesses'])

20

We have over 400 results, but the length is only 20? What's going on?

If you think about the results of a google search, we usually have MANY more results that Google will display at one time.  When we scroll down to the bottom of the results, we can see that Google has divided the results into several PAGES of results, instead of one extremely long page.

This is what the Yelp Fusion is doing as well! The general term for this is "Pagination".

The Yelp API will only return a "page" of 20 results at a time.

If we want to get the next page of results, we will perform another API call, but we will add an additional argument called "offset".
- The offset is what # result to use as the FIRST result for the page.
- If we had 20 businesses in our first result, we would want to add an offset of 20. 

In [11]:
# add offset to our original api call
search_results = yelp_api.search_query(location='Baltimore, MD,21202',
                                       term='Crab Cakes',
                                       offset = 20)

Now check the results

In [12]:
biz20 = pd.DataFrame(search_results['businesses'])
biz20.head(2)

id                                 alias  \
0  VLGdENTqjFJuCjg1SbW7_g            nicks-fish-house-baltimore   
1  ReNdW79g_1cxrgCJpFn7cA  b-and-o-american-brasserie-baltimore   

                     name                                          image_url  \
0        Nicks Fish House  https://s3-media3.fl.yelpcdn.com/bphoto/xMvMJ9...   
1  B&O American Brasserie  https://s3-media2.fl.yelpcdn.com/bphoto/STUupD...   

   is_closed                                                url  review_count  \
0      False  https://www.yelp.com/biz/nicks-fish-house-balt...           840   
1      False  https://www.yelp.com/biz/b-and-o-american-bras...           577   

                                          categories  rating  \
0         [{'alias': 'seafood', 'title': 'Seafood'}]     4.0   
1  [{'alias': 'newamerican', 'title': 'American (...     4.0   

                                         coordinates        transactions  \
0     {'latitude': 39.26036, 'longitude': -76.61421}          [delivery]   
1  {'latitude': 39.2899436853933, 'longitude': -7...  [delivery, pickup]   

  price                                           location         phone  \
0    $$  {'address1': '2600 Insulator Dr', 'address2': ...  +14103474123   
1    $$  {'address1': '2 N Charles St', 'address2': '',...  +14436926172   

    display_phone     distance  
0  (410) 347-4123  3984.496569  
1  (443) 692-6172   958.642226

You should have different results than your original call when you make the offset call.

Now you can combine the results (so far) into one dataframe using pd.concat

In [13]:
## concatenate the previous results and new results. 
businesses = pd.concat([biz, biz20],
                      ignore_index=True)
display(businesses)

id                                              alias  \
0   D9A33FM394q99o4QtK5YwA                       faidleys-seafood-baltimore-3   
1   ieS_5zqxDHcWMCm8BKUYbg               thames-street-oyster-house-baltimore   
2   u65W69AhbjUlvJJBkEhGNQ                     miss-shirleys-cafe-baltimore-9   
3   6am8TZAFnvND52MOz-Yctg                  mamas-on-the-half-shell-baltimore   
4   p_XPpHiZbdOccx2jFUgZJA                         the-local-oyster-baltimore   
5   fdo4pqfSYjQyW5_-rElfew                      faidley-s-seafood-baltimore-2   
6   BhnCaqbJtWrC36eVxlewSw                             l-p-steamers-baltimore   
7   1C8QSqc-erkkyAUywh501A                           angies-seafood-baltimore   
8   UplfpgvI5BfuDjtolDrmsA                          blackwall-hitch-baltimore   
9   NQu0GgFMcAB6ucmhSNpcSQ                         phillips-seafood-baltimore   
10  AjMBJPjnaGIVhBd08gpjvA  the-water-street-tavern-and-key-west-patio-bar...   
11  sUzGTfR81JKDBZlABjgtMQ            masons-famous-lobster-rolls-baltimore-3   
12  fyJgGCwQGr2lUjHrl9EygQ                          not-just-crab-baltimore-3   
13  ISn7WyGQaFpsVSRSh0NSqg                             sal-and-sons-baltimore   
14  SyLoKmaYwq9xCYeI6S6A3g                                fishnet-baltimore-2   
15  O2ZrDUuqXoluq9cmrFhoyg               mount-vernon-marketplace-baltimore-2   
16  7lDhtcolamUZw6oP5V1ehw                           flash-crabcake-baltimore   
17  s0LPuaTzR9euWzNT52o49Q               sugarbees-cafe-and-grill-baltimore-5   
18  CQ79htRPD6yxi0Zg4sYTNw                     the-capital-grille-baltimore-3   
19  n3bY9l_uDccV9X2TxH4rrg                        baltimore-seafood-baltimore   
20  VLGdENTqjFJuCjg1SbW7_g                         nicks-fish-house-baltimore   
21  ReNdW79g_1cxrgCJpFn7cA               b-and-o-american-brasserie-baltimore   
22  Kh1XJP54vra02w7vS5rRhg                        clock-restoration-baltimore   
23  d8hR3Vr27uoYMXaK5PIW0w                                 loch-bar-baltimore   
24  L9AsG9KRoRBBSbeTpNh3Jw                            rusty-scupper-baltimore   
25  6Omm4GMPqsuvpr7-oJwbow                       the-urban-oyster-baltimore-4   
26  TPQmNAM2g1vmm2xCfhyEJw                   pratt-street-ale-house-baltimore   
27  a2nFJQ12xNbRTm0__-hKdA        mos-fishermans-wharf-inner-harbor-baltimore   
28  g428REEbAhb4Ngumk5bJ6g                         lexington-market-baltimore   
29  heDO_cYM7C-BXmh17sMR-Q                   avenue-kitchen-and-bar-baltimore   
30  sdNaP8ecfDXzoigTjO2oTg                           blue-moon-cafe-baltimore   
31  8-cBsSXmLaGK6qCD0tmcZg                          the-food-market-baltimore   
32  GOvVNYe2DwjGzwdVJIKw1Q                    wicked-sisters-tavern-baltimore   
33  t0EA_fXewpQWfzWMrpNjIw                        peters-pour-house-baltimore   
34  Hpnp8R6mjwiSSNeCWBh-EQ                    david-and-dads-cafe-baltimore-4   
35  yITYeW3JZS557DIXj9Ivhg                             mick-o-sheas-baltimore   
36  uP4pCpjzUGMDIRzMtHDCkg                          old-bay-seafood-baltimore   
37  EHLjwy5Lcwoc0y0T8-K6hA                                topside-baltimore-2   
38  6ulR8s_SK3X0wYVQNZ0DSQ                          next-phaze-cafe-baltimore   
39  u4vKwc4p4HY2lT0Lneaz0w                         one-eyed-mikes-baltimore-2   

                                            name  \
0                              Faidley's Seafood   
1                     Thames Street Oyster House   
2                            Miss Shirley's Cafe   
3                       Mama's On The Half Shell   
4                               The Local Oyster   
5                              Faidley’s Seafood   
6                                   L P Steamers   
7                                Angie's Seafood   
8                                Blackwall Hitch   
9                               Phillips Seafood   
10  The Water Street Tavern & Key West Patio Bar   
11                  Mason's Famous Lobster Rolls   
12                        